src/mock_crust/support.rs

```diff
@@ -0,0 +1,449 @@
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+// It seems that code used only in tests is considered unused by rust.
+#![allow(unused)]
+
+use rand;
+use std::cell::{Cell, RefCell};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::io;
+use std::rc::{Rc, Weak};
+
+use super::crust::{ConnectionInfoResult, CrustEventSender, Event, OurConnectionInfo, PeerId,
+                   TheirConnectionInfo};
+
+/// Mock network. Create one before testing with mocks. Use it to create `Device`s.
+#[derive(Clone)]
+pub struct Network(Rc<RefCell<NetworkImpl>>);
+
+pub struct NetworkImpl {
+    services: HashMap<Endpoint, Weak<RefCell<ServiceImpl>>>,
+    next_endpoint: usize,
+    queue: VecDeque<(Endpoint, Endpoint, Packet)>,
+}
+
+impl Network {
+    /// Create new mock Network.
+    pub fn new() -> Self {
+        Network(Rc::new(RefCell::new(NetworkImpl {
+            services: HashMap::new(),
+            next_endpoint: 0,
+            queue: VecDeque::new(),
+        })))
+    }
+
+    /// Create new Device.
+    pub fn new_device(&self, config: Option<Config>, endpoint: Option<Endpoint>) -> Device {
+        let config = config.unwrap_or_else(|| Config::new());
+        let endpoint = endpoint.unwrap_or_else(|| self.gen_endpoint());
+
+        let device = Device::new(self.clone(), config, endpoint);
+        let _ = self.0.borrow_mut().services
+                                   .insert(endpoint, Rc::downgrade(&device.0));
+
+        device
+    }
+
+    pub fn gen_endpoint(&self) -> Endpoint {
+        let mut imp = self.0.borrow_mut();
+        let num = imp.next_endpoint;
+        imp.next_endpoint += 1;
+
+        Endpoint(num)
+    }
+
+    pub fn poll(&self) {
+        while let Some((sender, receiver, packet)) = self.pop_packet() {
+            self.process_packet(sender, receiver, packet);
+        }
+    }
+
+    fn send(&self, sender: Endpoint, receiver: Endpoint, packet: Packet) {
+        self.0.borrow_mut().queue.push_back((sender, receiver, packet));
+    }
+
+    fn pop_packet(&self) -> Option<(Endpoint, Endpoint, Packet)> {
+        self.0.borrow_mut().queue.pop_front()
+    }
+
+    fn process_packet(&self, sender: Endpoint, receiver: Endpoint, packet: Packet) {
+        if let Some(service) = self.find_service(receiver) {
+            service.borrow_mut().receive_packet(sender, packet);
+        } else {
+            // Packet was sent to a non-existing receiver.
+            if let Some(failure) = packet.to_failure() {
+                self.send(receiver, sender, failure);
+            }
+        }
+    }
+
+    fn find_service(&self, endpoint: Endpoint) -> Option<Rc<RefCell<ServiceImpl>>> {
+        self.0.borrow().services.get(&endpoint).and_then(|s| s.upgrade())
+    }
+}
+
+/// Device represents a mock version of a real machine connected to the network.
```

Comment:
That would probably be better - but still not completely accurate, as applications may start several `crust::Service` instances (and safe_launcher does, I think). Maybe just `ServiceHandle` or `ServiceRef`?